File: iosApp/iosAppTests/Utils/FormattedAlertTests.swift
//
//  FormattedAlertTests.swift
//  iosApp
//
//  Created by esimon on 4/21/26.
//  Copyright © 2026 MBTA. All rights reserved.
//

import Shared
import SwiftUI
import ViewInspector
import XCTest

@testable import iosApp

final class FormattedAlertTests: XCTestCase {
    func testTripShuttle() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripShuttleAlertSummary(
                tripIdentity: TripShuttleAlertSummary.SingleTrip(
                    tripTime: .init(
                        year: 2026,
                        month: .march,
                        day: 9,
                        hour: 12,
                        minute: 13,
                        second: 0
                    ),
                    routeType: .commuterRail,
                    fromStop: nil
                ),
                startStopName: "Oak Grove",
                endStopName: "Forest Hills",
                recurrence: nil
            )
        )
        XCTAssertEqual(
            "Shuttle buses replace the 12:13\u{202F}PM train from Oak Grove to Forest Hills",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testDownstreamTripShuttle() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripShuttleAlertSummary(
                tripIdentity: TripShuttleAlertSummary.SingleTrip(
                    tripTime: .init(
                        year: 2026,
                        month: .march,
                        day: 9,
                        hour: 12,
                        minute: 13,
                        second: 0
                    ),
                    routeType: .commuterRail,
                    fromStop: "Oak Grove"
                ),
                startStopName: "Ruggles",
                endStopName: "Forest Hills",
                recurrence: nil
            )
        )
        XCTAssertEqual(
            "12:13\u{202F}PM train from Oak Grove is replaced by shuttle buses from Ruggles to Forest Hills",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testThisTripShuttle() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripShuttleAlertSummary(
                tripIdentity: TripShuttleAlertSummary.ThisTrip(
                    routeType: .commuterRail
                ),
                startStopName: "Porter",
                endStopName: "North Station",
                recurrence: nil
            )
        )
        XCTAssertEqual(
            "Shuttle buses replace this train from Porter to North Station",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testDownstreamTripShuttleRecurrence() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripShuttleAlertSummary(
                tripIdentity: TripShuttleAlertSummary.SingleTrip(
                    tripTime: .init(
                        year: 2026,
                        month: .march,
                        day: 9,
                        hour: 12,
                        minute: 13,
                        second: 0
                    ),
                    routeType: .commuterRail,
                    fromStop: "Oak Grove"
                ),
                startStopName: "Ruggles",
                endStopName: "Forest Hills",
                recurrence: AlertSummary.RecurrenceDaily(
                    ending: AlertSummary.TimeframeThisWeek(
                        time: .init(
                            year: 2026,
                            month: .march,
                            day: 12,
                            hour: 9,
                            minute: 6,
                            second: 0
                        )
                    )
                )
            )
        )
        XCTAssertEqual(
            "12:13\u{202F}PM train from Oak Grove is replaced by shuttle buses from Ruggles to Forest Hills daily until Thursday",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testThisTripShuttleRecurrence() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripShuttleAlertSummary(
                tripIdentity: TripShuttleAlertSummary.ThisTrip(
                    routeType: .commuterRail
                ),
                startStopName: "Porter",
                endStopName: "North Station",
                recurrence: AlertSummary.RecurrenceDaily(
                    ending: AlertSummary.TimeframeThisWeek(
                        time: .init(
                            year: 2026,
                            month: .march,
                            day: 12,
                            hour: 9,
                            minute: 6,
                            second: 0
                        )
                    )
                )
            )
        )
        XCTAssertEqual(
            "Shuttle buses replace this train from Porter to North Station daily until Thursday",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testTripSuspension() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripSpecificAlertSummary(
                tripIdentity: TripSpecificAlertSummary.TripFrom(
                    tripTime: .init(
                        year: 2026,
                        month: .march,
                        day: 9,
                        hour: 12,
                        minute: 13,
                        second: 0
                    ),
                    routeType: .commuterRail,
                    stopName: "Ruggles"
                ),
                effect: .suspension,
                effectStops: nil,
                isToday: true,
                cause: .weather
            )
        )
        XCTAssertEqual(
            "12:13\u{202F}PM train from Ruggles is suspended today due to weather",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testThisTripSuspension() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripSpecificAlertSummary(
                tripIdentity: TripSpecificAlertSummary.ThisTrip(
                    routeType: .commuterRail
                ),
                effect: .suspension,
                effectStops: nil,
                isToday: true,
                cause: .weather
            )
        )
        XCTAssertEqual(
            "This train is suspended today due to weather",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testDownstreamTripSuspension() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripSpecificAlertSummary(
                tripIdentity: TripSpecificAlertSummary.TripFrom(
                    tripTime: .init(
                        year: 2026,
                        month: .march,
                        day: 9,
                        hour: 12,
                        minute: 13,
                        second: 0
                    ),
                    routeType: .commuterRail,
                    stopName: "Ruggles"
                ),
                effect: .suspension,
                effectStops: ["South Station"],
                isToday: true,
                cause: .weather
            )
        )
        XCTAssertEqual(
            "12:13\u{202F}PM train from Ruggles will terminate at South Station today due to weather",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testThisDownstreamTripSuspension() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripSpecificAlertSummary(
                tripIdentity: TripSpecificAlertSummary.ThisTrip(
                    routeType: .commuterRail
                ),
                effect: .suspension,
                effectStops: ["South Station"],
                isToday: true,
                cause: .weather
            )
        )
        XCTAssertEqual(
            "This train will terminate at South Station today due to weather",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testTripStopSkipped() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripSpecificAlertSummary(
                tripIdentity: TripSpecificAlertSummary.TripTo(
                    tripTime: .init(
                        year: 2026,
                        month: .march,
                        day: 9,
                        hour: 12,
                        minute: 13,
                        second: 0
                    ),
                    routeType: .commuterRail,
                    headsign: "South Station"
                ),
                effect: .stationClosure,
                effectStops: ["Back Bay", "Ruggles"],
                isToday: true,
                cause: .weather
            )
        )
        XCTAssertEqual(
            "12:13\u{202F}PM train to South Station will not stop at Back Bay and Ruggles today due to weather",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }

    func testThisTripStopSkipped() throws {
        let formatted = FormattedAlert(
            alert: nil,
            alertSummary: TripSpecificAlertSummary(
                tripIdentity: TripSpecificAlertSummary.ThisTrip(
                    routeType: .ferry
                ),
                effect: .dockClosure,
                effectStops: ["Rowes Wharf"],
                isToday: true,
                cause: .weather
            )
        )

        XCTAssertEqual(
            "This ferry will not stop at Rowes Wharf today due to weather",
            String(formatted.alertCardMajorBody.characters[...])
        )
    }
}

File: androidApp/src/androidTest/java/com/mbta/tid/mbta_app/android/util/FormattedAlertTests.kt
package com.mbta.tid.mbta_app.android.util

import androidx.compose.ui.platform.LocalResources
import androidx.compose.ui.test.junit4.createComposeRule
import com.mbta.tid.mbta_app.android.assertMatches
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.AlertSummary
import com.mbta.tid.mbta_app.model.AlertSummaryEntity
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.TripShuttleAlertSummary
import com.mbta.tid.mbta_app.model.TripSpecificAlertSummary
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import kotlin.test.assertEquals
import kotlin.time.Duration.Companion.hours
import kotlinx.coroutines.test.runTest
import kotlinx.datetime.Month
import org.junit.Rule
import org.junit.Test

class FormattedAlertTests {
    @get:Rule val composeTestRule = createComposeRule()

    @Test
    fun testTripSpecificShuttle() = runTest {
        val now = EasternTimeInstant(2026, Month.APRIL, 10, 15, 0, 0)
        val summary =
            TripShuttleAlertSummary(
                TripShuttleAlertSummary.SingleTrip(
                    now.plus(1.hours),
                    RouteType.COMMUTER_RAIL,
                    null,
                ),
                "Porter",
                "North Station",
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "Shuttle buses replace the **4:00\u202FPM** train from **Porter** to **North Station**",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex(
                    "Shuttle buses replace the 4:00\\sPM train from Porter to North Station",
                    RegexOption.IGNORE_CASE,
                ),
                summaryString,
            )
        }
    }

    @Test
    fun testTripSpecificDownstreamShuttle() = runTest {
        val now = EasternTimeInstant(2026, Month.APRIL, 10, 15, 0, 0)
        val summary =
            TripShuttleAlertSummary(
                TripShuttleAlertSummary.SingleTrip(
                    now.plus(1.hours),
                    RouteType.COMMUTER_RAIL,
                    "Concord",
                ),
                "Porter",
                "North Station",
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "**4:00\u202FPM** train from **Concord** is replaced by shuttle buses from **Porter** to **North Station**",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex(
                    "4:00\\sPM train from Concord is replaced by shuttle buses from Porter to North Station",
                    RegexOption.IGNORE_CASE,
                ),
                summaryString,
            )
        }
    }

    @Test
    fun testThisTripSpecificShuttle() = runTest {
        val summary =
            TripShuttleAlertSummary(
                TripShuttleAlertSummary.ThisTrip(RouteType.COMMUTER_RAIL),
                "Porter",
                "North Station",
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "Shuttle buses replace this train from **Porter** to **North Station**",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertEquals(
                "Shuttle buses replace this train from Porter to North Station",
                summaryString,
            )
        }
    }

    @Test
    fun testTripSpecificSuspension() = runTest {
        val now = EasternTimeInstant(2026, Month.APRIL, 10, 12, 13, 0)
        val summary =
            TripSpecificAlertSummary(
                TripSpecificAlertSummary.TripFrom(now, RouteType.COMMUTER_RAIL, "Ruggles"),
                Alert.Effect.Suspension,
                null,
                true,
                Alert.Cause.Weather,
            )
        val alert = ObjectCollectionBuilder.Single.alert { cause = Alert.Cause.Weather }
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "**12:13\u202FPM** train from **Ruggles** is suspended today due to weather",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex(
                    "12:13\\sPM train from Ruggles is suspended today due to weather",
                    RegexOption.IGNORE_CASE,
                ),
                summaryString,
            )
        }
    }

    @Test
    fun testTripSpecificDownstreamSuspension() = runTest {
        val now = EasternTimeInstant(2026, Month.APRIL, 10, 11, 15, 0)
        val summary =
            TripSpecificAlertSummary(
                TripSpecificAlertSummary.TripFrom(now, RouteType.COMMUTER_RAIL, "Concord"),
                Alert.Effect.Suspension,
                listOf("Porter"),
                true,
                Alert.Cause.Weather,
            )
        val alert = ObjectCollectionBuilder.Single.alert { cause = Alert.Cause.Weather }
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "**11:15\u202FAM** train from **Concord** will terminate at Porter today due to weather",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex(
                    "11:15\\s[Aa][Mm] train from Concord will terminate at Porter today due to weather"
                ),
                summaryString,
            )
        }
    }

    @Test
    fun testThisTripSpecificSuspension() = runTest {
        val summary =
            TripSpecificAlertSummary(
                TripSpecificAlertSummary.ThisTrip(RouteType.COMMUTER_RAIL),
                Alert.Effect.Suspension,
                null,
                true,
                Alert.Cause.Weather,
            )
        val alert = ObjectCollectionBuilder.Single.alert { cause = Alert.Cause.Weather }
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "This train is suspended today due to weather",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertEquals("This train is suspended today due to weather", summaryString)
        }
    }

    @Test
    fun testThisTripSpecificDownstreamSuspension() = runTest {
        val summary =
            TripSpecificAlertSummary(
                TripSpecificAlertSummary.ThisTrip(RouteType.COMMUTER_RAIL),
                Alert.Effect.Suspension,
                listOf("Porter"),
                true,
                Alert.Cause.Weather,
            )
        val alert = ObjectCollectionBuilder.Single.alert { cause = Alert.Cause.Weather }
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "This train will terminate at Porter today due to weather",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertEquals("This train will terminate at Porter today due to weather", summaryString)
        }
    }

    @Test
    fun testTripSpecificDownstreamStopClosure() = runTest {
        val now = EasternTimeInstant(2026, Month.APRIL, 10, 12, 13, 0)
        val summary =
            TripSpecificAlertSummary(
                TripSpecificAlertSummary.TripTo(now, RouteType.COMMUTER_RAIL, "Stoughton"),
                Alert.Effect.StopClosure,
                listOf("Back Bay", "Ruggles"),
                true,
                Alert.Cause.Weather,
            )
        val alert = ObjectCollectionBuilder.Single.alert { cause = Alert.Cause.Weather }
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "**12:13\u202FPM** train to **Stoughton** will not stop at **Back Bay** and **Ruggles** today due to weather",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex(
                    "12:13\\sPM train to Stoughton will not stop at Back Bay and Ruggles today due to weather",
                    RegexOption.IGNORE_CASE,
                ),
                summaryString,
            )
        }
    }

    @Test
    fun testThisTripSpecificStopClosure() = runTest {
        val summary =
            TripSpecificAlertSummary(
                TripSpecificAlertSummary.ThisTrip(RouteType.COMMUTER_RAIL),
                Alert.Effect.StationClosure,
                listOf("Porter"),
                true,
                Alert.Cause.Weather,
            )
        val alert = ObjectCollectionBuilder.Single.alert { cause = Alert.Cause.Weather }
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "This train will not stop at **Porter** today due to weather",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertEquals("This train will not stop at Porter today due to weather", summaryString)
        }
    }

    @Test
    fun testOneStopSkipped() = runTest {
        val summary =
            AlertSummary.Standard(
                Alert.Effect.StationClosure,
                AlertSummary.Location.AffectedStops(listOf("Back Bay")),
                AlertSummary.Timeframe.UntilFurtherNotice,
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "Trains will not stop at **Back Bay** until further notice",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex("Trains will not stop at Back Bay until further notice"),
                summaryString,
            )
        }
    }

    @Test
    fun testTwoStopsSkipped() = runTest {
        val summary =
            AlertSummary.Standard(
                Alert.Effect.StationClosure,
                AlertSummary.Location.AffectedStops(listOf("Back Bay", "Ruggles")),
                AlertSummary.Timeframe.UntilFurtherNotice,
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "Trains will not stop at **Back Bay** and **Ruggles** until further notice",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex("Trains will not stop at Back Bay and Ruggles until further notice"),
                summaryString,
            )
        }
    }

    @Test
    fun testThreeStopsSkipped() = runTest {
        val summary =
            AlertSummary.Standard(
                Alert.Effect.StationClosure,
                AlertSummary.Location.AffectedStops(listOf("Back Bay", "Ruggles", "Hyde Park")),
                AlertSummary.Timeframe.UntilFurtherNotice,
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "Trains will not stop at **Back Bay**, **Ruggles**, and **Hyde Park** until further notice",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex(
                    "Trains will not stop at Back Bay, Ruggles,? and Hyde Park until further notice"
                ),
                summaryString,
            )
        }
    }

    @Test
    fun testMultipleStopsSkipped() = runTest {
        val summary =
            AlertSummary.Standard(
                Alert.Effect.StationClosure,
                AlertSummary.Location.AffectedStops(
                    listOf("Back Bay", "Ruggles", "Hyde Park", "Readville")
                ),
                AlertSummary.Timeframe.UntilFurtherNotice,
            )
        val alert = ObjectCollectionBuilder.Single.alert()
        val format =
            FormattedAlert(
                alert,
                summary,
                AlertSummaryEntity(
                    null,
                    null,
                    null,
                    null,
                    "Trains will not stop at **multiple stops** until further notice",
                ),
            )
        composeTestRule.setContent {
            val summaryString = format.alertCardMajorBody(LocalResources.current).toString()
            assertMatches(
                Regex("Trains will not stop at multiple stops until further notice"),
                summaryString,
            )
        }
    }
}
